docs/docs/guides/05-type-system/01-philosophy.md

```diff
@@ -0,0 +1,79 @@
+---
+title: Philosophy
+tags:
+  - typescript
+  - type system
+  - type
+  - philosophy
+  - principles
+---
+
+### Context
+
+Since its release, Strapi's strength has laid in its **flexibility** and **extensibility**.
+
+It can handle a wide array of content and customizations, from content-types to plugins and extensions, most of them are defined dynamically in user applications.
+
+With TypeScript imposing itself as a standard in the JS ecosystem over the years, it became critical to develop a uniform system that streamlines this heavily customized content management's developer experience.
+
+This is in this context that a type system becomes an essential tool as it provides a single source of truth for standard types, like schema, attributes, and entities, whether they've been statically or dynamically created.
+
+On top of this, it offers a wide range of utilities to manipulate those types, and creates a unified, typed developer experience across the whole codebase.
+
+### Application 
+
+#### For Contributors
+
+One of its key advantages is the plethora of utilities it provides, which simplifies the manipulation of complex data structures, including content types, documents, and UIDs.
+
+Another significant advantage is the consistency it enforces within the codebase. This is achieved by defining shared types once and reusing them throughout the codebase.
+
+Moreover, the type system can effectively handle high levels of customization, thus allowing developers to focus their efforts on feature development and not TypeScript headaches.
+
+#### For Users
+
+Although Strapi users don't often engage directly with the type system, it nonetheless profoundly influences their Strapi experience.
+
+It ensures the provision of type-safe APIs (such as the document service), which greatly reduces runtime errors, offers a simple way of discovering APIs, and makes the system more reliable overall.
+
+Moreover, it arms users with ready-made types, which they can use to type-proof their applications or customizations.
+
+Lastly, the type system also provides users with type-manipulation utilities.
+
+### Challenges
+
+The main challenge faced by the type system is being able to maintain a high-quality TypeScript developer experience for different contexts, each with their singularities.
+
+#### For Contributors
+
+Contributors expects to create and/or use APIs that handle very generic data structures in the context of Strapi internals, that become context-aware and strongly typed when used by a user withing their application.
+
+#### For Users
+
+Users expects to have a TypeScript experience tailored around their own application, with their content-types, components, and plugins being strongly typed and recognized by the Strapi APIs.
+
+:::warning
+Creating such experiences without slowing down developers (both contributors and users) is a huge undertaking and should be considered as a top priority everytime the type system is modified. 
```

Comment:
Do we want to mention being careful to avoid breaking changes as well?

Reply:
We yet have to decide what our stance will be on this topic. We had a short discussion about this with @alexandrebodin but nothing definitive

Reply:
Although in any case, it's best to avoid breaking changes, we need to decide whether we'll completely forbid them or just avoid having too many

Reply:
I like the statement from react-query on non-breaking/breaking types 

https://tanstack.com/query/v4/docs/framework/react/typescript

We can't follow it to the later as most things are generated but sth to consider exploring

Reply:
I mentionned it to Piwi for our official doc